polkadot/node/core/dispute-coordinator/src/scraping/tests.rs

```diff
@@ -662,3 +662,276 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
+			result.contains(&(block_number, block_hash)),
+			"Expected to find ({}, {}) in the result",
+			block_number,
+			block_hash
+		);
+	}
+}
+
+#[test]
+fn inclusions_iterative_removal() {
+	let mut inclusions = Inclusions::new();
+
+	// Insert 3 candidates in the specified blocks
+	let candidate1 = make_candidate_receipt(BlakeTwo256::hash(&"c1".encode())).hash();
+	let candidate2 = make_candidate_receipt(BlakeTwo256::hash(&"c2".encode())).hash();
+	let candidate3 = make_candidate_receipt(BlakeTwo256::hash(&"c3".encode())).hash();
+	let candidate4 = make_candidate_receipt(BlakeTwo256::hash(&"c4".encode())).hash();
+	let candidate5 = make_candidate_receipt(BlakeTwo256::hash(&"c5".encode())).hash();
+
+	//	B	0	1	2	3
+	// C1	x
+	// C2		x
+	// C3	x		x
+	// C4			x
+	// C5				x
+	inclusions.insert(candidate1, 0, get_block_number_hash(0));
+	inclusions.insert(candidate2, 1, get_block_number_hash(1));
+	inclusions.insert(candidate3, 0, get_block_number_hash(0));
+	inclusions.insert(candidate3, 2, get_block_number_hash(2));
+	inclusions.insert(candidate4, 2, get_block_number_hash(2));
+	inclusions.insert(candidate5, 3, get_block_number_hash(3));
+
+	inclusions.remove_up_to_height(&0);
+	assert!(inclusions.contains(&candidate1), "Expected candidate1 to remain");
+	assert!(inclusions.contains(&candidate2), "Expected candidate2 to remain");
+	assert!(inclusions.contains(&candidate3), "Expected candidate3 to remain");
+	assert!(inclusions.contains(&candidate4), "Expected candidate4 to remain");
+	assert!(inclusions.contains(&candidate5), "Expected candidate5 to remain");
+
+	inclusions.remove_up_to_height(&1);
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(inclusions.contains(&candidate2), "Expected candidate2 to remain");
+	assert!(inclusions.contains(&candidate3), "Expected candidate3 to remain");
+	assert!(inclusions.contains(&candidate4), "Expected candidate4 to remain");
+	assert!(inclusions.contains(&candidate5), "Expected candidate5 to remain");
+
+	inclusions.remove_up_to_height(&2);
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(!inclusions.contains(&candidate2), "Expected candidate2 to be removed");
+	assert!(inclusions.contains(&candidate3), "Expected candidate3 to remain");
+	assert!(inclusions.contains(&candidate4), "Expected candidate4 to remain");
+	assert!(inclusions.contains(&candidate5), "Expected candidate5 to remain");
+
+	inclusions.remove_up_to_height(&10);
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(!inclusions.contains(&candidate2), "Expected candidate2 to be removed");
+	assert!(!inclusions.contains(&candidate3), "Expected candidate3 to be removed");
+	assert!(!inclusions.contains(&candidate4), "Expected candidate4 to be removed");
+	assert!(!inclusions.contains(&candidate5), "Expected candidate5 to be removed");
+}
+
+#[test]
+fn inclusions_duplicate_insertion_same_height_and_block() {
```

Comment:
I would be more interested if pruning works as expected for same height/same block/different block. As this was precisely what went wrong in the Kusama incident. 

Ideally we would also test whether pruning still works as desired if the same candidate was included at different heights.